E 32 defer level loading

bdaenen opened the pull request:
This should fix #32. I resorted to putting a `UIOverlay` inside the `ig.Preloader` as I did not know how to handle the spawning process of the preloader on each level. It also seems more logical, the preloader does far more than a UI element should do.

bdaenen committed:
Added an XHR helper for easy cross-browser get requests.

bdaenen committed:
Added and implemented a preloader so only the first level has to be initially loaded. See #32.

collinhover commented on lib/plusplus/core/game.js:
```diff
@@ -494,12 +512,49 @@ ig.module(
 
                         this.spawnEntity(ent.type, ent.x, ent.y, ent.settings);
 
+                        if ( ent.type === 'EntityLevelchange' ) {
```
This check won't work if the entity extends ig.EntityLevelChange. ig.game.spawnEntity returns the entity created, so this should probably be an instanceof check of the returned value instead.

bdaenen replied:
I'm actually really having some silly, unexpected issues with this. Why doesn't `ent instanceof ig.EntityLevelChange` ever return `true`? I've tried logging them separately and end up with an `EntityLevelChange` object and the basic `ig.Class` constructor. Am I missing something here :/?

collinhover replied:
`ent` is the data from the editor, so it is probably just a plain object. But the name of the variable is confusing, so maybe we can rename it to `entData`? 

The `ig.game.spawnEntity` method returns the actual entity that was created and spawned. Use that value in the instanceof check.

bdaenen replied:
That sounds incredibly logical :) It's definitely time to call it a day after the next push :)

collinhover commented on lib/plusplus/core/preloader.js:
```diff
@@ -0,0 +1,248 @@
+            createTransitioner: function() {
+
+                this.transitioner = ig.game.spawnEntity(ig.UIOverlay, 0, 0, {
+                    layerName: _c.TRANSITIONER_LAYER,
```
These config values should probably be moved into a new config namespace, such as "ig.CONFIG.PRELOADER". Maybe this could also have various config settings such as whether the preloader is enabled, etc.

collinhover replied:
edit: disregard part of the above comment. I still think a preloader config namespace is necessary, but the level transitioner should be handled by the game and not the preloader.

collinhover commented on lib/plusplus/core/game.js:
```diff
@@ -625,16 +680,16 @@ ig.module(
                     // complete transition
 
-                    if ( this.transitioner ) {
+                    if ( this.preloader.transitioner ) {
```
I think its good that you moved the transitioner to the preloader, but could we move all the transitioner code (spawning, fading, etc) into preloader methods? Then the game can just call a single method on the preloader and let the preloader handle the specifics.

collinhover replied:
edit: on second thought, maybe it would be best if the game handled level transitions. Can we let the preloader only do preloading, and the game handle building levels, tearing down levels, and transitioning between levels?

collinhover commented:
Great work, by the way =)

bdaenen commented:
I'll do some updating tomorrow and refactor the transitions :) shouldn't be too big of an issue at all! Thanks for the input ;)

collinhover commented:
Off topic: I wanted to mention your contributions at onGameStart 2013 in Warsaw, but I don't know how to pronounce your name. Can you help me out?

bdaenen commented:
Vassildador, or Benno Daenen :D?

bdaenen committed:
Moved transitioner back to ig.game.
Fixed an incorrect typeof check where instanceof should be used.
Added a config to enable/disable the preloader.

collinhover commented:
Looks ready to merge, let me give it a try!

collinhover commented:
When using the preloader, it does not seem to preload required modules, only images. I'll look into it, but let me know if I'm missing something!

bdaenen commented:
I'm afraid it only does levels with their related images so far. To extend it to sounds, modules and entities a far more complex structure is required >.< I don't think there's a reliable way to figure out when to load what

collinhover commented:
I've been running a lot of different tests, and I think you're right about the complex structure. To do something like that, we'd need to dispatch a signal when a module is loaded... I did set that up and got it running, but it required a pretty big change to the library and some ugly hacks to ImpactJS.

But that isn't the biggest issue =( if we use the preloader, we can't safely bake and minify our games, because some levels might require entities, modules, sounds, etc that the first level doesn't. Any thoughts on what we should do?

collinhover commented:
Hmm, I'd like to keep the preloader in the library, so what if we move the preloader from a core module to a helper module for now? The game won't try to preload, but the option will be there so we can still expand it if we figure out a good solution to the current issues.

bdaenen commented:
Sounds okay to me :) in a big game it'll still significantly decrease the initial load anyway so its of some use ^^

There should be a way to solve these issues though... the folks over at http://www.cross-code.com/ have it figured out although I think they heavily modified the impact engine itself :)
